### DISCUSSION TITLE
Is making a persisting request in a repository considered bad practice? Are repositories restricted solely to "SELECT" queries?

### DISCUSSION THREAD

**Discussion by hugo-fasone** [symfony/symfony#54243] [Q&A] [answered]
Hi,

I'm pretty new to Symfony and I've not seen any official recommendation about this

Actually I ask these question because of an underlying problem : I want to have a function that insert an Entity into the database (and do some other thing). I basically want something like this

```php

function add($params1, $params2 /*,$param3 ... */) {
    // Do some other things
    $myEntity = new MyEntity();
    $myEntity->setParams1($params1);
    // Do some other things
    $this->em->persist($myEntity);
    $this->em->flush();
    // Do some other things
}
```
Sort of a wrapper for persisting an entity but also doing other processes at the same time

Yes I'm aware of the listener/subscriber events, but I have my reasons for doing it with a function this way. My real dilemma is : where do i put this function? Inside a service or inside a repository? What is considered a good practice?

I've seen some responses for similar questions to be like "Just do it in your controller", the thing is i want this code to be reusable, and it's not only persisting an entity but also doing other things

For now I've put this function in a service, but i'd like to see the point of view of more exprienced Symfony devs

Thank you for your time

**Comment by MatTheCat** [comment, marked as answer]
Hello,

> Is making a persisting request in a repository considered bad practice?

Quite the opposite: a repository should provide a collection-like API for the objects it manages, which means allowing to add (amongst others) such objects.

> I have my reasons for doing it with a function this way.

It is weird to ask about good practices when you justify yourself using anti-patterns. Did you hear about the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) e.g.? What reason could force you to ignore this principle?

**Reply by hugo-fasone** [reply]
> Quite the opposite: a repository should provide a collection-like API for the objects it manages, which means allowing to add (amongst others) such objects.

Perfect thank you, that's what i was looking for

> It is weird to ask about good practices when you justify yourself using anti-patterns. Did you hear about the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) e.g.? What reason could force you to ignore this principle?

I think I know what you mean. So for the sake of it let's say that i want a function whose it's ONLY purpose (so that it doesn't violate the principle if I'm not mistaken) is to persist a certain entity to the database like so

```php
function add(string $title, string $content /*,$param3 ... */) {

    $post = new Post();
    $post->setTitle($title);
    $post->setContent($content);
    // ...
    $this->em->persist($post);
    $this->em->flush();
}
```

My question would be "where do i put this function, into a custom service or into the entity's repository?"

**Reply by MatTheCat** [reply]
You probably have an [anemic domain model](https://martinfowler.com/bliki/AnemicDomainModel.html), which would explain why you think it is a repository’s job to create entities. Do you see how it still violates the SRP?

A typical repository implementation would get an entity as an argument. Using the Doctrine ORM, this could be written as

```php
function add(Post $post): void
{
    $this->em->persist($post);
    $this->em->flush();
}
```

But if you don’t need entities, then you don’t need the ORM!

**Reply by hugo-fasone** [reply]
> You probably have an [anemic domain model](https://martinfowler.com/bliki/AnemicDomainModel.html), which would explain why you think it is a repository’s job to create entities. Do you see how it still violates the SRP?

Mhh yes I think see what you mean.
But in this case, with this code
```php
$post = new Post();
$post->setTitle($title);
$post->setContent($content);
// ...
$this->repo->add($post);
```
Can i still replace it by this code and still not violate that principle?
```php
$post = $this->repo->hydrate($title, $content)
$this->repo->add($post);
```
Or with something like this?
```php
$post = new Post()
$post = $this->repo->hydrate($post, $title, $content)
$this->repo->add($post);
```

The reason why i'm insisting so much is that, first of all I'm trying to understand these principles, and second of all it seems "weird" to me to call like 20 or 30 setter functions every time i need to add an Entity when i could just call a function with all the parameters that "prepare" the entity, it seems "nicer" that way (even though i know "nicer" is not an justification, but i don't get why calling all those 30 setter functions everytime would be good either)

EDIT : I guess the "hydrate" approach is still an "anemic domain model" anti-pattern, so i guess it's a no go

Thank you for your time

**Reply by MatTheCat** [reply]
I understand you want to abstract the entity creation (we’re not talking about persistence here). The associated pattern is called “**factory**”. Whether it’s an entity’s constructor, another entity’s method or a dedicated service, its goal is to provide a new entity instance. E.g. if your domain model says “an author writes posts”, then you may have

```php
class Author
{
    public function writePost(string $title, string $content): Post
    {
        // …
    }
}
```

To persist a new post, you’d then write

```php
$post = $author->writePost($title, $content);
$postRepository->add($post);
```

This illustrates neatly the separation between creation and persistence: these are different concepts so they’re not merged into a single service.

---

Now, the example above suggests you do have a domain model, which does not seem to be the case. If you use entities as simple database row classes, then you in fact does not need entities. E.g. what about

```php
class PostRepository
{
    function add(array $data): void
    {
        $this->conn->insert('posts', $data);
    }
}

$postRepository->add(['title' => $title, 'content' => $content]);
```

This code is way simpler because it does not need an object as intermediate representation (that is, a [DTO](https://en.wikipedia.org/wiki/Data_transfer_object)). If your entities are nothing more than that, it may be appropriate not to use an ORM.
